lib/chef/resource/macos_userdefaults.rb

```diff
@@ -73,54 +107,90 @@ def coerce_booleans(val)
+        # todo: This is a pretty basic implementation. PRs welcome ;)
+        case state.stdout[0]
+        when "{" # dict aka hash
+          # https://rubular.com/r/cBnFu1nttMdsXq
+          data = /^\s{2,}(.*)\s=\s(.*);/.match(state.stdout)
+          fail_if_unparsable(data, state.stdout)
+          value Hash[*data.captures]
+        when "(" # array
+          # https://rubular.com/r/TfYejXUJny11OG
+          data = /^\s{2,}(.*),?/.match(state.stdout)
+          fail_if_unparsable(data, state.stdout)
+          value data.captures
+        else # a string/int/float/bool
+          value state.stdout.strip
+        end
+      end
+
+      #
+      # If there were not matches raise a warning that we couldn't parse the output of the defaults
+      # CLI and return a nil current_resource by calling current_value_does_not_exist!
+      #
+      # @param [MatchData] match_data
+      # @param [String] stdout
+      #
+      def fail_if_unparsable(match_data, stdout)
+        return unless match_data.captures.empty?
+
+        Chef::Log.warn("Could not parse macos defaults CLI value data: #{stdout}.")
+        current_value_does_not_exist!
       end
 
       action :write do
-        description "Write the setting to the specified domain"
+        description "Write the value to the specified domain/key."
 
-        unless current_resource.is_set
-          cmd = ["defaults write"]
-          cmd.unshift("sudo") if new_resource.sudo
+        converge_if_changed do
+          cmd = defaults_write_cmd
+          Chef::Log.debug("Updating defaults value by shelling out: #{cmd.join(" ")}")
 
-          cmd << if new_resource.global
-                   "NSGlobalDomain"
-                 else
-                   "'#{new_resource.domain}'"
-                 end
+          if new_resource.user.nil?
+            shell_out!(cmd)
+          else
+            shell_out!(cmd, user: new_resource.user)
+          end
+        end
+      end
 
-          cmd << "'#{new_resource.key}'" if new_resource.key
+      action_class do
+        def defaults_write_cmd
           value = new_resource.value
-          type = new_resource.type.empty? ? value_type(value) : new_resource.type
-          # creates a string of Key1 Value1 Key2 Value2...
-          value = value.map { |k, v| "\"#{k}\" \"#{v}\"" }.join(" ") if type == "dict"
-          if type == "array"
-            value = value.join("' '")
-            value = "'#{value}'"
-          end
+          type = new_resource.type || value_type(value)
+
+          value = case type
+                  when "dict"
+                    # creates an array of Key1 Value1 Key2 Value2...
+                    value.flatten.map { |x| "'#{x}'" }
+                  when "array"
+                    value.map { |x| "'#{x}'" }
+                  else
+                    "'#{value}'"
+                  end
+
+          cmd = ["defaults", "write", "'#{new_resource.domain}'", "'#{new_resource.key}'"]
```

Comment:
the single-quotes shouldn't be needed there if you use the command as an array:

```
[phil@rider ~]$ cat /tmp/test.sh
#!/bin/bash

echo "\$1 is $1"
[phil@rider ~]$ /tmp/test.sh foo bar
$1 is foo
```
and:
```
irb(main):010:0> cmd = ['/tmp/test.sh', 'foo bar', 'baz']
irb(main):011:0> puts Mixlib::ShellOut.new(cmd).run_command.stdout
$1 is foo bar
```

Reply:
We need the quotes to supports keys with a space. You can pass whatever you want into defaults as long as you single quote it on the CLI:

If I try to chef-apply something like this with a space in the key:

```ruby
macos_userdefaults "value with space" do
  domain "/Library/Preferences/ManagedInstalls"
  key "Log File"
  value "/Library/Managed Installs/Logs/ManagedSoftwareUpdate.log"
end
```

With the quoting code removed:

```
❰tsmith❙~/dev/work/chef(git≠macos_userdefaults)❱✔≻ bundle exec chef-apply test.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * macos_userdefaults[value with space] action write

    ================================================================================
    Error executing action `write` on resource 'macos_userdefaults[value with space]'
    ================================================================================

    Mixlib::ShellOut::ShellCommandFailed
    ------------------------------------
    Expected process to exit with [0], but received '1'
    ---- Begin output of ["defaults", "write", "/Library/Preferences/ManagedInstalls", "Log File", "'/Library/Managed Installs/Logs/ManagedSoftwareUpdate.log'"] ----
    STDOUT:
    STDERR: 2020-07-14 21:31:09.951 defaults[55022:8620703] Could not write domain /Library/Preferences/ManagedInstalls; exiting
    ---- End output of ["defaults", "write", "/Library/Preferences/ManagedInstalls", "Log File", "'/Library/Managed Installs/Logs/ManagedSoftwareUpdate.log'"] ----
    Ran ["defaults", "write", "/Library/Preferences/ManagedInstalls", "Log File", "'/Library/Managed Installs/Logs/ManagedSoftwareUpdate.log'"] returned 1

    Resource Declaration:
    ---------------------
    # In test.rb

      1: macos_userdefaults "value with space" do
      2:   domain "/Library/Preferences/ManagedInstalls"
      3:   key "Log File"
      4:   value "/Library/Managed Installs/Logs/ManagedSoftwareUpdate.log"
      5: end

    Compiled Resource:
    ------------------
    # Declared in test.rb:1:in `run_chef_recipe'

    macos_userdefaults("value with space") do
      action [:write]
      default_guard_interpreter :default
      declared_type :macos_userdefaults
      cookbook_name "(chef-apply cookbook)"
      recipe_name "(chef-apply recipe)"
      domain "/Library/Preferences/ManagedInstalls"
      key "Log File"
      value "/Library/Managed Installs/Logs/ManagedSoftwareUpdate.log"
    end

    System Info:
    ------------
    chef_version=16.3.8
    platform=mac_os_x
    platform_version=10.15.6
    ruby=ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-darwin19]
    program_name=/usr/local/lib/ruby/gems/2.7.0/bin/chef-apply
    executable=/usr/local/Cellar/ruby/2.7.1_2/bin/bundle

[2020-07-14T21:31:09-07:00] FATAL: Stacktrace dumped to /Users/tsmith/.chef/cache/chef-stacktrace.out
[2020-07-14T21:31:09-07:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2020-07-14T21:31:09-07:00] FATAL: Mixlib::ShellOut::ShellCommandFailed: macos_userdefaults[value with space] ((chef-apply cookbook)::(chef-apply recipe) line 1) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '1'
---- Begin output of ["defaults", "write", "/Library/Preferences/ManagedInstalls", "Log File", "'/Library/Managed Installs/Logs/ManagedSoftwareUpdate.log'"] ----
STDOUT:
STDERR: 2020-07-14 21:31:09.951 defaults[55022:8620703] Could not write domain /Library/Preferences/ManagedInstalls; exiting
---- End output of ["defaults", "write", "/Library/Preferences/ManagedInstalls", "Log File", "'/Library/Managed Installs/Logs/ManagedSoftwareUpdate.log'"] ----
Ran ["defaults", "write", "/Library/Preferences/ManagedInstalls", "Log File", "'/Library/Managed Installs/Logs/ManagedSoftwareUpdate.log'"] returned 1
```

With the code as it is in this branch:

```
❰tsmith❙~/dev/work/chef(git✱≠macos_userdefaults)❱✘≻ bundle exec chef-apply test.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * macos_userdefaults[value with space] action write
    - update value with space
    -   set value to "/Library/Managed Installs/Logs/ManagedSoftwareUpdate.log" (was {"Identifier"=>"HUBaseColorLightBlue"})
```

Reply:
dafuq. that's... not how `exec()` is supposed to work. Goddamnit Apple, what have you done.

Reply:
We're trying to run this kind of thing on the CLI:

```
defaults write '/Library/Preferences/ManagedInstalls' 'Log File' -dict 'Identifier' 'HUBaseColorLightBlue'
```

Reply:
yes, but it shouldn't, unless a shell is interpreting it. When you use Mixlib::ShellOut with an array, it's calling `exec()` and the second argument is an array and the array has each argument as a single element in it - regardless of spaces. To put that in ruby it'd be `exec('defaults', ['write', '/Library/Preferences/ManagedInstalls' 'Log File', ...])` - in which case there's nothing to misunderstand that space.

The only tine that should be a problem is if it's being passed to a shell first, which it shouldn't be if Mixlib::ShellOut is called with an array instead of a string, which you are doing.

I can't tell if this is some weird bug in Mixlib on Mac, or some other crazy thing Mac is doing. Unless I'm misunderstanding something, which is always possible. @lamont-granquist might have thoughts here... am I crazy?

Reply:
Everything is converted to array format and the single quotes nuked. Value was the thing holding it up.